### ISSUE TITLE
Inconsistent peer pool API

### ISSUE THREAD

**Issue by chapko** [thaliproject/Thali_CordovaPlugin#1813]
`ThaliPeerPoolInterface` has `enqueue` method which either returns `null` or throws an error. But all the implementations (`ThaliPeerPoolDefault` and `ThaliPeerPoolOneAtATime`) return errors instead. Also `ThaliNotificationClient` does not wrap `enqueue` calls in the `try-catch` block, which can lead to the unhandled errors.

My suggestion is to stick to the `ThaliPeerPoolInterface` design and always throw errors. It can break custom peer pool implementations, but it is easy to update by just replacing `return` with `throw`.

**Comment by yaronyg** [comment]
But isn't the bug that enqueue throws an error? Shouldn't it return an error instead? Throwing in a heavily asynchronous environment like our Node code is a recipe for nightmares. If we ever throw anywhere but in the immediate call thread then we can't be sure who will get the throw. That's why returning errors is usually preferred. It's cleaner and the control flow is the same, promises or not.

**Comment by chapko** [comment]
I disagree. Synchronous functions return value on success and throw error on error. Async callback-based functions do it by passing null or error as the first argument to the callback. Promise-based functions return promise, which is eventually either rejected with error or fulfilled with a value.

Everything else is just misleading.

> Throwing in a heavily asynchronous environment like our Node code is a recipe for nightmares

It is easy to solve with promises because they transform thrown errors into rejected promises by design.

So in my opinion the question here is whether we consider unsuccessful enqueuing worth throwing an error. Alternatively we can just return true/false from enqueue if we are not interested in error itself.

I've found only 2 places where we enqueue actions:
```JavaScript
// thaliNotificationClient.js
var enqueueError = this._thaliPeerPoolInterface.enqueue(action);

if (!enqueueError) {
  this.peerDictionary.addUpdateEntry(peer, peerEntry);
} else {
  logger.warn('_createAndEnqueueAction: failed to enqueue an item: %s',
    enqueueError.message);
}
```

```JavaScript
// thaliPullReplicationFromNotification.js
var enqueueError = this._thaliPeerPoolInterface.enqueue(newAction);
if (enqueueError) {
  logger.warn('_peerAdvertisesDataForUsHandler: failed to enqueue an item: %s',
    enqueueError.message);
}
```

And in both cases errors are just logged. This is basically the same as returning true/false and logging errors from peerPoolInterface itself.

More node way would probably be emitting 'error' event from notification client (and from pull replication). And if nobody subscribed to this event then just throw an error. In this case unexpected errors won't be unnoticed. But I don't know yet how this error event should be propagated to the top of the stack where user could properly handle it. Also it requires our own implementation to never cause errors under normal conditions, which is not the case right now (it can throw errors when we stop everything).

To summarize it all, I have 2 suggestions:
1. Always return value or throw errors form `enqueue`, transform it to 'error' event and make user handle it. We have to make sure these errors are caused only by incorrect custom peer pool implementation.
2. Return true/false from enqueue. It clearly states that enqueueing invalid action is not something exceptional and we do not care much about it.

**Comment by yaronyg** [comment]
I'm wrong. You're right. I think throwing makes sense as enqueuing bad things is, bad. :)
